### PR TITLE
[FLINK-26249][table-planner] Run BuiltInFunctionsTestBase and BuiltInAggregateFunctionsTestBase in parallel

### PR DESCRIPTION
**Pull request by slinkydeveloper**
This PR includes two commits to port `BuiltInFunctionsTestBase` and `BuiltInAggregateFunctionsTestBase` to JUnit 5 and run those tests in parallel.

On my machine running the whole `functions` package and subpackage tests with non-parallel execution takes 55 secs, while with parallel it takes around ~25 secs

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
@flinkbot run azure

**Comment by slinkydeveloper** [comment]
Run 3 times in my local CI, all green:

* https://dev.azure.com/francescoguard/Flink/_build/results?buildId=764&view=results
* https://dev.azure.com/francescoguard/Flink/_build/results?buildId=773&view=results
* https://dev.azure.com/francescoguard/Flink/_build/results?buildId=774&view=results

**Comment by slinkydeveloper** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java [hunk 1]
```diff
@@ -60,127 +62,68 @@
...
-                    new MiniClusterResourceConfiguration.Builder()
-                            .setNumberTaskManagers(1)
-                            .setNumberSlotsPerTaskManager(1)
-                            .build());
+    Configuration getConfiguration() {
+        return new Configuration();
+    }
 
-    @Parameter public TestSpec testSpec;
+    abstract Stream<TestSetSpec> getTestCaseSpecs();
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:83]: `getTestSetSpecs `

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java [hunk 2]
```diff
@@ -342,18 +286,53 @@ TestSpec testResult(
             return this;
         }
 
+        Stream<TestCase> getTestCases(Configuration configuration) {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:289]: private?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java [hunk 3]
```diff
@@ -342,18 +286,53 @@ TestSpec testResult(
...
+            return testItems.stream().map(testItem -> getTestCase(configuration, testItem));
+        }
+
+        private TestCase getTestCase(Configuration configuration, TestItem testItem) {
+            return new TestCase(
+                    testItem.toString(),
+                    () -> {
+                        final TableEnvironmentInternal env =
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:297]: I don't remember but is there an actual reason for `TableEnvironmentInternal` otherwise let's expose only `TableEnvironment`.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java [hunk 4]
```diff
@@ -342,18 +286,53 @@ TestSpec testResult(
...
         @Override
         public String toString() {
             return (definition != null ? definition.getName() : "Expression")
                     + (description != null ? " : " + description : "");
         }
     }
 
-    private interface TestItem {
-        // marker interface
+    interface TestItem {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:331]: why not private? also below

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CoalesceFunctionITCase.java [hunk 5]
```diff
@@ -20,23 +20,20 @@
...
 /** Test {@link BuiltInFunctionDefinitions#COALESCE} and its return type. */
-public class CoalesceFunctionITCase extends BuiltInFunctionTestBase {
+class CoalesceFunctionITCase extends BuiltInFunctionTestBase {
 
-    @Parameterized.Parameters(name = "{index}: {0}")
-    public static List<TestSpec> testData() {
-        return Collections.singletonList(
-                TestSpec.forFunction(BuiltInFunctionDefinitions.COALESCE)
+    @Override
+    public Stream<TestSetSpec> getTestCaseSpecs() {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CoalesceFunctionITCase.java:34]: could also be default scoped?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/FieldAccessFromTableITCase.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+import static org.apache.flink.table.api.DataTypes.ROW;
+import static org.apache.flink.table.api.DataTypes.STRING;
+import static org.apache.flink.table.api.Expressions.$;
+
+/**
+ * Regular tests. See also {@link ConstructedAccessFunctionsITCase} for tests that access a nested
+ * field of an expression or for {@link BuiltInFunctionDefinitions#FLATTEN} which produces multiple
+ * columns from a single one.
+ */
+class FieldAccessFromTableITCase extends BuiltInFunctionTestBase {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/FieldAccessFromTableITCase.java:40]: What is the current status of the MiniCluster? Will it be reused across tests nowadays? Because in the past the guideline was to not create too many classes as the cluster init is expensive. This is why summarized functions into one class in the past.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/FieldAccessFromTableITCase.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+                                ARRAY(BIGINT().notNull()).notNull(),
+                                MAP(STRING(), BIGINT().notNull()).nullable(),
+                                MAP(STRING(), BIGINT().notNull()).notNull(),
+                                ROW(FIELD("nested", BIGINT().notNull())).nullable(),
+                                ROW(FIELD("nested", BIGINT().notNull())).notNull())
+                        // accessing elements of MAP or ARRAY is a runtime operations,
+                        // we do not know about the size or contents during the inference
+                        // therefore the results are always nullable
+                        .testResult(
+                                resultSpec($("f0").at(1), "f0[1]", null, BIGINT().nullable()),
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/FieldAccessFromTableITCase.java:85]: remove the outer `resultSpec(...)`, `testResult` can take it directly. also in other tests.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java [hunk 8]
```diff
@@ -72,7 +71,8 @@
  * Tests for catalog and system functions in a table environment.
  *
  * <p>Note: This class is meant for testing the core function support. Use {@link
- * BuiltInFunctionTestBase} for testing individual function implementations.
+ * org.apache.flink.table.planner.functions.BuiltInFunctionTestBase} for testing individual function
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java:74]: Undo this change

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the PR @slinkydeveloper. Great that we can speed up our tests with this. Big +1. I added some last comments.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java (hunk 3 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:297]: It's needed to access the catalog manager to manually run the type inference, check line 353 of this class

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/FieldAccessFromTableITCase.java (hunk 6 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/FieldAccessFromTableITCase.java:40]: With this re-org each class extending from `BuiltInFunctionTestBase` will have its own `MiniCluster`, and the same `MiniCluster` is reused across all test cases

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/FieldAccessFromTableITCase.java (hunk 7 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/FieldAccessFromTableITCase.java:85]: in this case i'm not sure it does. `testResult` is either a vararg of result spec or a function taking 5 args. I rather prefer to keep it as it is.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java (hunk 8 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java:74]: I can't, because of the changed visibility of BuiltInFunctionTestBase

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/FieldAccessFromTableITCase.java (hunk 7 above)
Comment (twalthr) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/FieldAccessFromTableITCase.java:85]: `BuiltInFunctionTestBase.TestSpec#testResult(org.apache.flink.table.expressions.Expression, java.lang.String, java.lang.Object, org.apache.flink.table.types.AbstractDataType<?>)` has 4 args as well.

**Review by twalthr — APPROVED** [review]
LGTM
